### PR TITLE
Qt Add GUI wif import dialog

### PR DESCRIPTION
**Pull request by chromatic**
Addresses #1808.

### CODE REVIEW

File: src/qt/bitcoingui.cpp
Comment (rnicoll) [inline src/qt/bitcoingui.cpp:397]:
```suggestion
    importPrivateKeyAction->setStatusTip(tr("Import a Dogecoin private key"));
```

File: src/qt/forms/importkeysdialog.ui
Comment (rnicoll) [inline src/qt/forms/importkeysdialog.ui:45]:
```suggestion
            <string>Private key to import into your wallet</string>
```

**Review by rnicoll — CHANGES_REQUESTED** [review]
Couple of very minor nit-picks, and I need to actually try running it!

File: src/qt/forms/importkeysdialog.ui
Comment (chromatic) [reply src/qt/forms/importkeysdialog.ui:45]: Addressed this too.

**Comment by rnicoll** [comment]
Main issue is this does a rescan on the GUI thread, so the whole interface hangs. For a rescan can you perhaps open a dialog that shows the rescan is in progress, but without freezing the whole UI? Could be a modal dialogue to protect against anyone triggering multiple rescans.

**Comment by chromatic** [comment]
> Main issue is this does a rescan on the GUI thread, so the whole interface hangs. For a rescan can you perhaps open a dialog that shows the rescan is in progress, but without freezing the whole UI? Could be a modal dialogue to protect against anyone triggering multiple rescans.

I'll look into that. The current behavior duplicates the console window behavior, so it may be worth extracting the key behavior into key code shared between these two import mechanisms. I don't know if that should or shouldn't be in scope of this PR though.

**Comment by rnicoll** [comment]
My inclination is get it working well, then clean up behind the scenes.

Let me know if you would like a hand with doing the rescan on a different thread, I haven't done Qt work in ages, but happy to take a look at least.

**Comment by chromatic** [comment]
Thanks! I'll take the approach you suggest.

I'll look into the thread work and follow up with you if I get stuck. If you or anyone else reading this can quickly point to code that does something similar, that might be enough to get me going in the right direction (but I wouldn't spend time digging for it; if it's there, I'll find it).

**Comment by chromatic** [comment]
@rnicoll I've just pushed code that accomplished this, though it's a little less elegant than I'd like. The happy paths seem to work okay in my testing.

**Review by Germanstreetstyle — APPROVED** [review]

**Review by Germanstreetstyle — APPROVED** [review]

**Review by Germanstreetstyle — APPROVED** [review]

**Comment by patricklodder** [comment]
closing & reopening for CI

**Comment by chromatic** [comment]
One of the CI failures looks like a missing BDB header:

```
In file included from ./wallet/walletdb.h:11:0,
2261                 from ./wallet/wallet.h:19,
2262                 from ./qt/importkeysdialog.h:7,
2263                 from qt/moc_importkeysdialog.cpp:8:
2264./wallet/db.h:21:20: fatal error: db_cxx.h: No such file or directory
2265 #include <db_cxx.h>
2266                    ^
```

The other failure (on Darwin) I believe I've fixed with the latest commit.

**Comment by patricklodder** [comment]
Re: bib failure at https://travis-ci.com/github/dogecoin/dogecoin/jobs/506711074#L2264 - this is because with `NO_WALLET=1` compilation, there is nowhere to store the keys.

I think the problem is the inclusion of `wallet/wallet.h` in `importkeysdialog.h`? I think you can just import that in the .cpp file instead and it be good, since you already have `walletmodel.h` in there?

**Comment by chromatic** [comment]
I like the idea, but it might not be that easy because `importkeysdialog.h` declares `rescanWallet(CWallet*, CBlockIndex*)` as a `Q_SIGNAL`.

I've mentioned both classes in the header file directly. It'd be nice to exclude all of this QT wallet importing code in case of `NO_WALLET=1` but that's not obvious to me right now.

File: src/qt/bitcoingui.cpp
Comment (patricklodder) [inline src/qt/bitcoingui.cpp:396]: Non-blocking: Would love to have a new icon for this. Can merge regardless now though, we can change it later too.

File: src/qt/forms/importkeysdialog.ui
Comment (patricklodder) [inline src/qt/forms/importkeysdialog.ui:69]: This tooltip needs to be changed.

File: src/qt/forms/importkeysdialog.ui
Comment (patricklodder) [inline src/qt/forms/importkeysdialog.ui:56]: This tooltip needs to be changed.

File: src/qt/importkeysdialog.cpp
Comment (patricklodder) [inline src/qt/importkeysdialog.cpp:126]: Are we sure that when a key already exists, we want to react differently than when it doesn't? IIRC this is why RPC doesn't throw an error on duplicate key import, because that could be used to expose which keys are in the wallet?

**Review by patricklodder — CHANGES_REQUESTED** [review]

File: src/qt/importkeysdialog.cpp
Comment (chromatic) [reply src/qt/importkeysdialog.cpp:126]: Yeah, I'm not sure what the right balance is here. It'd be nice to give some suggestion of what the user could do to correct the problem, but I tried to phrase things vaguely enough that it's not leaking information.

Maybe something more like

  Couldn't add private key. This can be for several reasons:
  
   * the key already exists
   * the key is malformed
   * the wallet is malformed
 
I don't care much for that; it's a lot of text, it's vague, and it still suffers from the flaw you mention. Maybe "Failed to add private key" is the least worst option.

Comment (patricklodder) [reply src/qt/importkeysdialog.cpp:126]: Alright. I'm okay with "Failed to add private key" to unblock this for now. 

I will manually test every scenario I can think of between RPC and UI before merge, if I find any useful improvements we can address that in a subsequent PR.

File: src/qt/forms/importkeysdialog.ui
Comment (patricklodder) [inline src/qt/forms/importkeysdialog.ui:10]: Height is a bit high at the moment.

File: src/qt/forms/importkeysdialog.ui
Comment (patricklodder) [inline src/qt/forms/importkeysdialog.ui:24]: I'd remove these, there's no need for the tab widget as there is only one view. Result:
Screen Shot 2021-05-24 at 02 11 27

File: src/qt/forms/importkeysdialog.ui
Comment (patricklodder) [inline src/qt/forms/importkeysdialog.ui:107]: matching widget closing tag for `QTabWidget`

File: src/qt/forms/importkeysdialog.ui
Comment (patricklodder) [inline src/qt/forms/importkeysdialog.ui:196]: Not sure if we need this spacer?

**Review by patricklodder — CHANGES_REQUESTED** [review]
I only have look & feel items left. Everything works well, thank you!!!

Tested on macOS 11.3

File: src/qt/forms/importkeysdialog.ui
Comment (chromatic) [reply src/qt/forms/importkeysdialog.ui]: I changed it to 200, because that's the smallest I could get it to go locally.

File: src/qt/forms/importkeysdialog.ui
Comment (chromatic) [reply src/qt/forms/importkeysdialog.ui:102]: Done, thanks!

File: src/qt/forms/importkeysdialog.ui
Comment (chromatic) [reply src/qt/forms/importkeysdialog.ui]: Looks like we don't. Good catch.

**Comment by patricklodder** [comment]
This is good for me now. Mind rebase-squashing your commits with `git rebase -i 1.14.4-dev` into one big commit? Then we're good to merge and have our first end-to-end UI improvement in.

Thank you for the work!

**Comment by chromatic** [comment]
Squashed and force pushed.

**Review by rnicoll — APPROVED** [review]
Built & tested importing of keys from a remote node, all works as expected in positive scenarios at least.

**Comment by rnicoll** [comment]
Also tested the simple invalid private key scenario. The other failure cases in theory should never occur, and I can't see how to test without mocking parts of the backend, seems extreme.

**Review by patricklodder — APPROVED** [review]
ACK! Thank you so much!!!

**Pull request merged**
